oak_attestation_verification/src/verifier.rs

```diff
@@ -85,15 +87,42 @@ pub fn verify_internal(
         (Some(endorsements::Type::Cb(ends)), Some(reference_values::Type::Cb(rvs))) => {
             verify_cb(evidence, ends, rvs)
         }
-        (None, None) => anyhow::bail!("Endorsements and reference values both empty"),
-        (None, Some(_)) => anyhow::bail!("Endorsements are empty"),
-        (Some(_), None) => anyhow::bail!("Reference values are empty"),
-        (Some(_), Some(_)) => anyhow::bail!("Mismatch between endorsements and reference values"),
+        (None, None) => Err(anyhow::Error::msg(
+            "endorsements and reference values both empty",
+        )),
+        (None, Some(_)) => Err(anyhow::Error::msg("endorsements are empty")),
+        (Some(_), None) => Err(anyhow::Error::msg("reference values are empty")),
+        (Some(_), Some(_)) => Err(anyhow::Error::msg(
+            "mismatch between endorsements and reference values",
+        )),
+    };
+    if per_setup_result.is_err() {
+        return failure(per_setup_result.err().unwrap());
+    }
+
+    let pkeys = chain_result.unwrap();
+    let encryption_cose_key = pkeys.get(0).unwrap();
+    let encryption_public_key = encryption_cose_key
+        .clone()
+        .to_vec()
+        .map_err(|_cose_err| anyhow::anyhow!("could not serialize encryption public key"));
+    if encryption_public_key.is_err() {
+        return failure(encryption_public_key.err().unwrap());
+    }
+    let signing_cose_key = pkeys.get(1).unwrap();
+    let signing_public_key = signing_cose_key
+        .clone()
+        .to_vec()
+        .map_err(|_cose_err| anyhow::anyhow!("could not serialize signing public key"));
+    if signing_public_key.is_err() {
+        return failure(signing_public_key.err().unwrap());
     }
+
+    success(encryption_public_key.unwrap(), signing_public_key.unwrap())
 }
 
 /// Verifies signatures along the certificate chain induced by DICE layers.
-fn verify_dice_chain(evidence: &Evidence) -> anyhow::Result<()> {
+fn verify_dice_chain(evidence: &Evidence) -> anyhow::Result<[CoseKey; 2]> {
```

Comment:
I don't think we should return an array here, but rather a defined struct with named fields. Also, we should convert the CoseKey struct into the actual public key types (one HPKE public key, one ECDSA signing keys) so that the result is explicit.